src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -107,30 +100,42 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
      *            until the `Cargo.toml` becomes sound
      */
     override val projectDescription: CargoProjectDescription?
-        get() = lock (lock) { cached }
+        get() {
+            check(ApplicationManager.getApplication().isReadAccessAllowed)
+            return cached
+        }
 
-    private fun notifyCargoProjectUpdate(r: UpdateResult) {
+    private fun commitUpdate(r: UpdateResult) {
         ApplicationManager.getApplication().assertIsDispatchThread()
+        if (module.isDisposed) return
+
+        runWriteAction {
+            if (r is UpdateResult.Ok) {
+                cached = r.projectDescription
+                updateModuleDependencies(r.projectDescription);
+            }
+            notifyCargoProjectUpdate(r)
```

Comment:
If notifying under _local_ lock isn't such a great problem, notifying holding a _global write lock_ is seeking for a trouble for sure: action inside listeners may accidentally trigger read-actions, and then we'd get a deadlock on a singe-thread.